tests/unit/acquisition/test_function.py

```diff
@@ -461,6 +471,175 @@ def prepare_acquisition_function(
+            2,
+            1.0,
+            0.01,
+            1e-2,
+            id="1d_input_2obj_gp_var_1",
+        ),
+        pytest.param(
+            1,
+            200_000,
+            tf.constant([[0.3, 0.2], [0.2, 0.22], [0.1, 0.25], [0.0, 0.3]]),
+            2,
+            2.0,
+            0.01,
+            1e-2,
+            id="1d_input_2obj_gp_var_2",
+        ),
+        pytest.param(
+            2, 50_000, tf.constant([[0.0, 0.0]]), 2, 1.0, 0.01, 1e-2, id="2d_input_2obj_gp_var_2"
+        ),
+        pytest.param(
+            3,
+            50_000,
+            tf.constant([[2.0, 1.0], [0.8, 3.0]]),
+            2,
+            1.0,
+            0.01,
+            1e-2,
+            id="3d_input_2obj_gp_var_1",
+        ),
+        pytest.param(
+            4,
+            100_000,
+            tf.constant([[3.0, 2.0, 1.0], [1.1, 2.0, 3.0]]),
+            3,
+            1.0,
+            0.01,
+            1e-2,
+            id="4d_input_3obj_gp_var_1",
+        ),
+    ],
+)
+def test_expected_hypervolume_improvement(
+    input_dim: int,
+    num_samples_per_point: int,
+    existing_observations: tf.Tensor,
+    obj_num: int,
+    variance_scale: float,
+    rtol: float,
+    atol: float,
+) -> None:
+    # Note: the test data number grows exponentially with num of obj
+    data_num_seg_per_dim = 2  # test data number per input dim
+    N = data_num_seg_per_dim ** input_dim
+    xs = tf.convert_to_tensor(
+        list(itertools.product(*[list(tf.linspace(-1, 1, data_num_seg_per_dim))] * input_dim))
+    )
+
+    xs = tf.cast(xs, dtype=existing_observations.dtype)
+    model = ModelStack(
+        *[(_mo_test_models[_](kernel_amplitude=variance_scale), 1) for _ in range(obj_num)]
+    )
+
+    mean, variance = model.predict(xs)
+
+    # [f_samples, B, L]
+    predict_samples = tfp.distributions.Normal(mean, tf.sqrt(variance)).sample(
+        num_samples_per_point
+    )
+    _pareto = Pareto(existing_observations)
+    ref_pt = get_reference_point(_pareto.front)
+    lb_points, ub_points = _pareto.hypercell_bounds(
+        tf.constant([-math.inf] * ref_pt.shape[-1]), ref_pt
+    )
+
+    # calc MC approx EHVI
+    splus_valid = tf.reduce_all(
+        tf.tile(ub_points[tf.newaxis, :, tf.newaxis, :], [num_samples_per_point, 1, N, 1])
+        > tf.expand_dims(predict_samples, axis=1),
+        axis=-1,
+    )  # num_cells x B
+    splus_idx = tf.expand_dims(tf.cast(splus_valid, dtype=ub_points.dtype), -1)
+    splus_lb = tf.tile(lb_points[tf.newaxis, :, tf.newaxis, :], [num_samples_per_point, 1, N, 1])
+    splus_lb = tf.maximum(splus_lb, tf.expand_dims(predict_samples, 1))
+    splus_ub = tf.tile(ub_points[tf.newaxis, :, tf.newaxis, :], [num_samples_per_point, 1, N, 1])
+    splus = tf.concat([splus_idx, splus_ub - splus_lb], axis=-1)
+
+    ehvi_approx = tf.transpose(tf.reduce_sum(tf.reduce_prod(splus, axis=-1), axis=1, keepdims=True))
+    ehvi_approx = tf.reduce_mean(ehvi_approx, axis=-1)
```

Comment:
Whats going on here?

Reply:
i.e. what is splus?

Reply:
Here is a bit dirty indeed, basically this is another way of calculating HVI, the code is mainly inheriting the [gpflowopt implementation](https://github.com/GPflow/GPflowOpt/blob/master/gpflowopt/acquisition/hvpoi.py) with a modification to allow MC based acq calculating.

The main idea here for calculating ehvi is that, for a certain (partitioned) cell: 

- `splus_valid`: denote if this predicting point (or `predict_samples` in this case) can contribute hypervolume improvement in the cell (it can only contribute if it can dominate the upper bound point (i.e., worst point) of the cell, which is what `>` and `reduce_all` does).
- `splus_ub`, variable storing the upper bound of the partitioned cell
- `splus_lb`, variable storing the dimensional-wise max(lower bound, predict point)
- Line 633: `splus_ub - splus_lb` is a way of representing the (dimensional-wise) improvement, which can be used to calculate the HVI in this cell
- Line 636: average through MC sample

- splus: a variable containing validity label: `splus_valid`, as well as the contribution part:`splus_ub - splus_lb`. Then we can make use of it to calculate the HVI given that it is a valide contribution.

A maybe more interpetable example can be: 
so consider a simplest case with only 1 cell (2obj): lb: [-inf, -inf],  ub: [0, 0], 
1. for pred [-0.5, -0.5], as it dominates [0, 0], `splus_valid` = 1 means it is a valid point (can contribute on this cell)
the `splus_lb` will record something like max([-inf, -inf], [-0.5, -0.5]), and then use the splus_ub - splus_lb can finally be used for calculating its hvi.
2. for pred [-0.5, 1], or [1, 1], `splus_valid`=0, means it does not contributed in this cell.

Reply:
Could you perhaps add a couple of inline comments about whats going on? 

Basically a comment for each of the bullet points above

Reply:
Then I think we can merge. Next we can do the synthetic function PR and then get a notebook in?

Reply:
Comments have been added.

Reply:
The synthetic function PR → notebook is a good order.